[Woo POS] Add celebration UX to cash payments

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
This PR adds _chaching_ sound + haptic to cash payment success, which so far we only do via card payments. The feedback/question section below regarding tests can be handled here as well or as a separate PR with more tests around the class.

## Feedback/question:
1. I wasn't sure the best place to add this, I went ahead with the OrderController because:
- AggregateModel feels it shouldn't need to care, and if we handle it here on cash success then we'd need to make the function public and call it from somewhere else, for really no reason.
- OrderService it's in Yosemite, while the PaymentCaptureCelebration is above in WooCommerce.
- We can keep the logic encapsulated in the OrderController and just trigger when the result is success, which seems adequate.

We can move it somewhere else if feels more appropriate 🤔  

2. Testing (we can add those here or separately, I found myself blocked by the tests):

The value of testing the celebration being triggered on success is minimal, but while doing so I found out that we don't really have tests around this bit of logic neither. I seem unable to run the test without crashing the app due the action dispatcher erroring out because it's not running in the Main Thread. I've tried several approaches, but more or less all of them are similar. Some examples:

```
Inject let mockStores = MockStoresManager(sessionManager: .testingInstance), then:

    @Test func collectCashPayment_when_successful_calls_celebrate() async throws {
        // Given
        let orderItem = OrderItem.fake()
        let fakeOrder = Order.fake().copy(items: [orderItem])
        mockOrderService.orderToReturn = fakeOrder
        await sut.syncOrder(for: [makeItem()], retryHandler: {})

        mockStores.whenReceivingAction(ofType: OrderAction.self) { action in
            switch action {
            case let .updateOrder(_, order, _, _, onCompletion):
                onCompletion(.success(order))
            default:
                #expect(Bool(false), "Unexpected action: \(action)")
            }
        }

        let completionResult = await withCheckedContinuation { continuation in
            Task { @MainActor in
                try? await sut.collectCashPayment()
                continuation.resume(returning: true)
            }
        }

        #expect(completionResult == true)
        #expect(mockPaymentCelebration.celebrationWasCalled == true)
}
```

```
@Test func collectCashPayment_when_successful_calls_celebrate() async throws {
        // Given
        let orderItem = OrderItem.fake()
        let fakeOrder = Order.fake().copy(items: [orderItem])
        mockOrderService.orderToReturn = fakeOrder
        await sut.syncOrder(for: [makeItem()], retryHandler: {})

        let completionResult: Bool = await withCheckedContinuation { continuation in
            mockStores.whenReceivingAction(ofType: OrderAction.self) { action in
                switch action {
                case let .updateOrder(_, order, _, _, onCompletion):
                    onCompletion(.success(order))
                default:
                    #expect(Bool(false), "Unexpected action \(action)")
                }
            }
            
            Task { @MainActor in
                try await sut.collectCashPayment()
            }
        }
        
        // Then
        #expect(completionResult == true)
        #expect(mockPaymentCelebration.celebrationWasCalled == true)
    }
```

Specially in the second case I'd expect the logic happens on main, but that's not the case. Any feedback is appreciated here.

## Testing
- In POS, create an pay an order via cash. Note how the "chaching" sounds on success.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
add celebration UX to cash payments

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented:
@iamgabrielma 

> Specially in the second case I'd expect the logic happen on main, but that's not the case. Any feedback is appreciated here.

If I understand the issue correctly, Swift Testing runs tests in a background rather than a main thread by default. You could add `@MainActor` attribute before a test name to try again.

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleOrderController.swift:
```diff
@@ -123,7 +130,11 @@ final class PointOfSaleOrderController: PointOfSaleOrderControllerProtocol {
+                                                 onCompletion: { [weak self] result in
+                guard let self = self else { return }
+                if case .success = result {
+                    self.celebrate()
```
I didn't know we had a celebration protocol 😀

I don't have strong objections but for me the sound is a UI/UX related, similarly to animation or confetti effects, therefore it's more appropriate for them to be placed closely to the view. Either in view itself or within the model. What do you think?

iamgabrielma replied:
> I didn't know we had a celebration protocol

It has some funny history behind.  When we ran unit tests locally you suddenly got hit with ~10 "chaching" sounds full-blast coming from your machine totally unexpectedly which wasn't pleasant 😂 Now this is disabled for dev builds

>  I don't have strong objections but for me the sound is a UI/UX related, similarly to animation or confetti effects, therefore it's more appropriate for them to be placed closely to the view. Either in view itself or within the model. What do you think

That makes sense, and sounds good, I guess even in the view/viewhelper itself it's fine to call the class and method directly as it's very encapsulated 🤔  I'll play a bit more with the tests before merging anything.

iamgabrielma committed:
Add test for celebration

We need to declare the mocked StoreManager within the test scope, otherwise will run on a different thread and when actions are dispatched to the store would crash

iamgabrielma committed:
lint

iamgabrielma commented:
> If I understand the issue correctly, Swift Testing runs tests in a background rather than a main thread by default. You could add @MainActor attribute before a test name to try again.

Thanks for the tip, I figured out the issue and added the test on https://github.com/woocommerce/woocommerce-ios/pull/15022/commits/82484f86d38a81ace341b991fdf3b67879b426a1 

Declaring the MockStoresManager on the file level is what was making the action dispatcher crash, it seems that was running on a different thread that the rest of test logic. Moving this to the test scope, declaring the local SUT injecting the store, and marking the test as MainActor solved the crash. After that it was not crashing anymore, but wasn't passing nor failing either and just stuck in `Testing WooCommerceTests` but never finished. The console shows that it was due to leaking its continuation, so I've added `continuation.resume...` after completion and ✅

staskus commented:
@iamgabrielma thanks for following up and sharing what the issue was!
